kcrypt/encryptor.go

```diff
@@ -0,0 +1,735 @@
+
+	if useRemoteKMS {
+		logger.Logger.Info().
+			Str("challenger_server", kcryptConfig.ChallengerServer).
+			Bool("mdns", kcryptConfig.MDNS).
+			Msg("Using remote KMS for encryption")
+		encryptor = &RemoteKMSEncryptor{
+			logger:       logger,
+			kcryptConfig: kcryptConfig,
+		}
+	} else if isUKI {
+		logger.Logger.Info().Msg("Using TPM with PCR policy for encryption (UKI mode)")
+		encryptor = &TPMWithPCREncryptor{
+			logger:         logger,
+			bindPublicPCRs: bindPublicPCRs,
+			bindPCRs:       bindPCRs,
+		}
+	} else {
+		logger.Logger.Info().Msg("Using local TPM NV passphrase for encryption")
+		encryptor = &LocalTPMNVEncryptor{
+			logger:       logger,
+			kcryptConfig: kcryptConfig,
+		}
+	}
+
+	if err := encryptor.Validate(); err != nil {
+		return nil, fmt.Errorf("encryptor validation failed: %w", err)
+	}
+
+	return encryptor, nil
+}
+
+// scanCollectorConfig scans for configuration and returns the collector config.
+func scanCollectorConfig(logger types.KairosLogger) *collector.Config {
+	o := &collector.Options{NoLogs: true, MergeBootCMDLine: true}
+	if err := o.Apply(collector.Directories(DefaultConfigDirs...)); err != nil {
+		logger.Debugf("scanCollectorConfig: error applying collector options: %v", err)
+		return nil
+	}
+
+	collectorConfig, err := collector.Scan(o, func(d []byte) ([]byte, error) {
+		return d, nil
+	})
+	if err != nil {
+		logger.Debugf("scanCollectorConfig: error scanning for config: %v", err)
+		return nil
+	}
+
+	return collectorConfig
+}
+
+// detectUKIMode detects if the system is running in UKI mode
+// This checks for the presence of UKI-specific indicators.
+func detectUKIMode(logger types.KairosLogger) bool {
+	// Check if we're booted from a UKI by looking for systemd UKI indicators
+	// The most reliable way is to check if /run/systemd/tpm2-pcr-signature.json exists
+	// This file is created by systemd when booting from a UKI with PCR signatures
+	if _, err := os.Stat("/run/systemd/tpm2-pcr-signature.json"); err == nil {
+		logger.Logger.Debug().Msg("Detected UKI mode: found /run/systemd/tpm2-pcr-signature.json")
+		return true
+	}
+
+	// Alternative: check for /run/systemd/tpm2-pcr-public-key.pem
+	if _, err := os.Stat("/run/systemd/tpm2-pcr-public-key.pem"); err == nil {
+		logger.Logger.Debug().Msg("Detected UKI mode: found /run/systemd/tpm2-pcr-public-key.pem")
+		return true
+	}
+
+	logger.Logger.Debug().Msg("Not running in UKI mode")
+	return false
+}
+
+// partitionInfo holds information about a partition for unlocking.
+type partitionInfo struct {
+	DevicePath    string
+	PartitionName string
+	Partition     *block.Partition
+}
+
+// Locked returns true if the partition is currently locked (encrypted and not unlocked).
+func (p *partitionInfo) Locked() bool {
+	if p == nil {
+		return false
+	}
+	return !utils.Exists(p.MapperPath())
+}
+
+// MapperPath returns the device mapper path for the partition (e.g., /dev/mapper/sda1).
+func (p *partitionInfo) MapperPath() string {
```

Comment:
why is this needed but the devicepath not? Isnt it possible to calculate both values at init of the partition info as to not have to call the method?